toolchain/lex/string_literal.cpp

```diff
@@ -488,6 +490,25 @@ auto StringLiteral::ComputeCharLiteralValue(
     emitter.Emit(text_.begin(), CharLiteralRaw);
   }
 
+  if (content_.starts_with("\\x")) {
+    CARBON_DIAGNOSTIC(CharLiteralHexEscape, Error,
+                      "escape sequence `\\x` in character literal");
+    emitter.Emit(text_.begin(), CharLiteralHexEscape);
+    return std::nullopt;
+  }
+
+  if (content_.size() == 1 &&
+      static_cast<unsigned char>(content_.front()) < 0x20) {
```

Comment:
Should 0x7F (DEL) also be disallowed here?

Reply:
Probably. #1964 says specifically:

> ASCII control codes (0…31), including whitespace characters other than word space (tab, line feed, carriage return, form feed, and vertical tab), except when specified with an escape sequence.

... but I think what we should actually do is to disallow all Unicode control characters ([character class Cc](https://www.compart.com/en/unicode/category/Cc)), which is U+0000 - U+001F and U+007F - U+009F.